Use fortranformat

LiamPattinson opened the pull request:
Instead of using utilities such as `ChunkOutput`, `write_1d`, `write_2d`, `f2s`, etc, uses [fortranformat](https://pypi.org/project/fortranformat/) to handle Fortran formatted IO.

- Reimplements `geqdsk.read` and `geqdsk.write`
- Specifies the G-EQDSK standard more rigorously
- Adds utility functions that can read/write ND Fortran arrays to file

Builds upon PR https://github.com/freegs-plasma/FreeQDSK/pull/3, so may need to be revised if that PR needs to be changed significantly. Makes progress towards resolving Issue https://github.com/freegs-plasma/FreeQDSK/issues/2. Similar to an [open PR to FreeGS](https://github.com/freegs-plasma/freegs/pull/80), although it uses `fortranformat` in a slightly different way (no generators, just uses a consistent file handle directly).

Haven't yet updated A-EQDSK, but it shouldn't take too long to push similar changes there.

LiamPattinson committed:
Add Sphinx docs, updated docs for _fileutils and geqdsk

LiamPattinson committed:
Add docs for A-EQDSK

LiamPattinson committed:
Remove dedent in aeqdsk.fields()

Found to reduce readability when calling help(fields)

LiamPattinson committed:
Rewritten geqdsk

- Uses fortranformat to assist with formatted IO
- Specifies G-EQDSK standard more rigorously.
- Provides generalised utility functions for reading/writing
  Fortran arrays.

LiamPattinson committed:
Rewritten A-EQDSK

- Uses fortranformat and read_array/write_array in place of older
  file utils
- Attempts to more rigorously define A-EQDSK files, though there's
  not a lot of documentation out there!
- Better support for older file types that don't have an extended
  section or are missing some values on the end (or even have
  extra ones we're not expecting)

LiamPattinson commented:
This update has now pretty comprehensively rewritten the library to use fortranformat and more rigorously define the file formats, and I'm planning on making a few more additions later today. I've also expanded the documentation beyond the previous PR. Should this PR take precendence?

Things I still need to finish up:
- Need testing data. Currently testing by writing to a string buffer, then reading info back from it. If both the read and write functions make the same mistake, there's no way for us to know. Any idea where I could get some real G-EQDSK/A-EQDSK files that can be freely used?
- Can add some further fields to the end of A-EQDSK
- AFAIK, the only current users are FreeGS and Pyrokinetics. FreeGS is using some of our private methods and in their `pyproject.toml` ask for  `freeqdsk >= 0.1`, so I can't remove unused utility functions without breaking FreeGS. It'd be nice if those utilities could be ported back over to FreeGS.

Note: I've tested using this build in place of 0.1 with Pyrokinetics and it passed all tests, so that's encouraging!

bendudson commented:
Thanks @LiamPattinson ! I know there was some discussion among the OMFIT folks about EQDSK formats and FreeGS. I'll try pinging them to see if they can help test this.

LiamPattinson committed:
Added example geqdsk and aeqdsk files, updated geqdsk tests

LiamPattinson commented:
- Added some G-EQDSK files in use by Pyrokinetics. Let me know if there's a licensing issue here and I'll revert.
- Added an A-EQDSK file provided by Koki Imada (don't know his GitHub handle!).
- Expanded on the G-EQDSK tests, though still need to add some failure cases.

@bendudson Thanks! I've been having trouble finding detailed standards for A-EQDSK files, and have mostly pieced things together from the FreeGS implementation, [eqtools](https://github.com/PSFCPlasmaTools/eqtools), and the old [BOUT implementation](https://github.com/bendudson/BOUT/blob/master/tools/tokamak_grids/pyGridGen/aeqdsk.py). I'm hoping to add p files soon, and information there is even more scarce. Getting more eyes on the problem will definitely be helpful, especially if it's from people who use the files regularly and have access to the Fortran that generates them.

bendudson commented:
My experience is that there is quite a lot of variation in real GEQDSK files in the wild. It would be good to have a writer that strictly followed the specification (whatever that is), and a reader that is quite forgiving and can cope with variations.

bclyons12 commented:
The rigorous standard for a gEQDSK is: 
 - 'QPSI' positive
 - 'PSIRZ' increases from axis to edge
 - sign of 'Ip' and 'Bt' are given with respect to a counter-clockwise toroidal angle

One way to think about this is that, depending on the sign of 'Ip' and 'Bt', this corresponds to four possible COCOS representation (1,3,5, and 7).
```
        All use sigma_RpZ=+1 (phi is counterclockwise) and exp_Bp=0 (psi is flux/2.*pi)
        We want
        sign(psi_edge-psi_axis) = sign(Ip)*sigma_Bp > 0  (psi always increases in gEQDSK)
        sign(q) = sign(Ip)*sign(Bt)*sigma_rhotp > 0      (q always positive in gEQDSK)
        ::
            ============================================
            Bt    Ip    sigma_Bp    sigma_rhotp    COCOS
            ============================================
            +1    +1       +1           +1           1
            +1    -1       -1           -1           3
            -1    +1       +1           -1           5
            -1    -1       -1           +1           7
```

LiamPattinson commented:
I've come across a few variations in units and coordinate systems in G-EQDSK files, but so far all of them I've seen follow the same formatting rules. Am I good to assume the data will be written using the `(5e16.9)` Fortran format, or will the `read` function need to fall back on a more generic regex?

bendudson commented:
@LiamPattinson I'm not certain, but I think it should be safe to assume that standard Fortran format.

LiamPattinson committed:
Add read_line and write_line, pass plain formats to read/write_array

Removes direct usage of fortranformat in geqdsk and aeqdsk

LiamPattinson committed:
Add fileutils tests, add geqdsk tests, scramble geqdsk test files

LiamPattinson committed:
Add tests for aeqdsk using real file

The file in use contains values that we currently aren't accounting
for at the end of the extended section.

LiamPattinson committed:
Add extra fields to aeqdsk

LiamPattinson committed:
aeqdsk bug fixes, expanded tests

LiamPattinson commented:
I've finished adding tests for this update. A-EQDSK now recognises some extra fields, but I've seen indications that the last few lines aren't properly standardised and they might mean different things depending on the source of the file.

@ZedThree Please could you take a look over this when you next have time?

ZedThree commented on pyproject.toml:
```diff
@@ -23,22 +24,29 @@ classifiers = [
+    "sphinx == 3.5.4",
+    "myst_parser == 0.14.0",
+    "jinja2 < 3.1.0",
+    "sphinx_rtd_theme",
```
I have a feeling these are just copy-pasted from pyrokinetics :)

The sphinx version is very old and doesn't work with python >= 3.10, and I think myst isn't needed for this project. These two should be changed at least.

This is personal preference, but I think `sphinx-book-theme` looks nicer than the RTD theme.

A couple of suggestions for other useful plugins:

```suggestion
     "sphinx >= 5.3",
     "sphinx-book-theme >= 0.4.0rc1",
     "sphinx-argparse-cli >= 1.10.0",
     "sphinx_autodoc_typehints >= 1.19",
```

ZedThree commented on tests/utils.py:
```diff
@@ -0,0 +1,39 @@
+
+def scramble_geqdsk(filename: Path, output: Path, magnitude: float = 0.1) -> None:
+    """
+    Used to redact G-EQDSK data, converting actual data into random numbers with the
```
Neat!

ZedThree commented on freeqdsk/_fileutils.py:
```diff
@@ -127,3 +190,174 @@ def next_value(fh):
+            # Read ND arrays as flattened 1D arrays, then reshape to requested shape
+            return read_array((np.prod(shape),), fh, fmt).reshape(shape, order="F")
+    else:
+        raise ValueError("'shape' should be a scalar or a 1D array")
```
Minor, but moving this block up and guarding on something like `if shape.ndim not in [0, 1]` means you could drop one level of indentation from the `shape.ndim == 1` block

ZedThree commented on freeqdsk/aeqdsk.py:
```diff
@@ -1,280 +1,582 @@
+            )
+        return int(data[field.name])
+    else:
+        return len(data.get(field.length_of, []))
```
Move the guard clause to the top:

```suggestion
    if field.name not in data:
        return len(data.get(field.length_of, []))

    # Check length of array, if present
    if field.length_of in data and data[field.name] != len(data[field.length_of]):
        raise ValueError(
            f"'{field.name}' should be the length of the array '{field.length_of}'"
        )
    return int(data[field.name])
```

ZedThree commented on freeqdsk/aeqdsk.py:
```diff
@@ -1,280 +1,582 @@
+            raise ValueError(f"'{field.name}' should have length '{field.has_length}'")
+        return result
+    else:
+        return np.zeros(data.get(field.has_length, 0))
```
Move guard clause to top

ZedThree commented on freeqdsk/aeqdsk.py:
```diff
@@ -294,26 +596,94 @@ def write(data, fh):
+            extended = True
+            break
+
+    if extended:
```
Could bail early and unindent next block?

```suggestion
    if not extended:
        return
```

ZedThree commented on freeqdsk/aeqdsk.py:
```diff
@@ -294,26 +596,94 @@ def write(data, fh):
-            else:
-                output.write(data.get(key, default))
+    # Output first block of general data
+    write_array([_float_field(field, data) for field in _general_block_1], fh, data_fmt)
```
Is it worth wrapping this pattern up into a local helper function? Or are there too many variations?

ZedThree commented on freeqdsk/aeqdsk.py:
```diff
@@ -335,27 +706,268 @@ def read(fh):
+    extended_values = read_array("all", fh, data_fmt)
+    if len(extended_values) > len(_extended_general):
+        warnings.warn(
+            "Encountered variables at the end of an A-EQDSK file that are not "
+            "recognised by FreeQDSK. Please consider contributing to the project "
+            "and letting us know what these are!"
```
Nice! Worth putting in the link to the repo here?

ZedThree commented on freeqdsk/aeqdsk.py:
```diff
@@ -335,27 +706,268 @@ def read(fh):
     return data
+
+
+# Dynamic docstring generation
```
This is very fancy! It's probably worth putting a short note as to _why_ the docstring is dynamically generated

ZedThree reviewed (approved):
LGTM, a few little comments. I would definitely bump sphinx version at the very least

But very nice, comprehensive tests and docs, thanks!

LiamPattinson committed:
Update docs

- Use sphinx book theme
- Use autodoc type hints

LiamPattinson committed:
Simplify read_array

LiamPattinson committed:
Simplify aeqdsk

LiamPattinson commented:
Thanks for the tips! I've implemented the things you suggested:

- Docs updated. autodoc typehints will make things easier going forwards. It seems it isn't handling `np.typing.ArrayLike` well, but it's only a minor issue.
- Simplified `read_array`. Rewrote it so that a scalar `shape` is the 'standard' path through the function, and all ND arrays go through the same process of reading a flattened array and reshaping.
- Simplified A-EQDSK. Now writes the last line instead of going up to the last field.
